uint8[] data
string hashstr